### PR TITLE
Include exit data in backtrace, print first 128 chars if valid utf8.

### PR DESCRIPTION
**Pull request by vyzo**
See https://github.com/filecoin-project/ref-fvm/issues/1210

### CODE REVIEW

**Comment by vyzo** [comment]
CI failure is "No space left on device"

**Comment by Stebalien** [comment]
@vyzo when that happens, head to https://github.com/filecoin-project/ref-fvm/actions/caches and start wrecking.

**Comment by Stebalien** [comment]
Looks like clippy doesn't like you.

**Comment by vyzo** [comment]
yeah, who would have thought that map/flatten combo is verbotten.

**Review by Stebalien — COMMENTED** [review]
Looks correct but can you test against a solidity contract to see what happens?

**Comment by vyzo** [comment]
yes, working on the test to verify correct behaviour.
So far I have verified that the error data are in the backtrace using fevm, now I have to figure out how to print it.

File: fvm/src/call_manager/backtrace.rs [hunk 1]
```diff
@@ -82,18 +86,45 @@ pub struct Frame {
...
-            Address::new_id(self.source),
-            self.method,
-            &self.message,
-            self.code,
-        )
+        if let Some(data) = &self.data {
+            let data_repr = if let Ok(BytesDe(bytes)) = data.deserialize() {
+                if let Ok(s) = String::from_utf8(bytes) {
+                    if s.len() > MAX_BACKTRACE_DATA_LEN {
+                        s[..MAX_BACKTRACE_DATA_LEN].to_owned() + "..."
```
[the remark this answers is not in the document]
Comment (vyzo) [reply fvm/src/call_manager/backtrace.rs:99]: ah, ok, looks nice.

File: fvm/src/call_manager/backtrace.rs [hunk 2]
```diff
@@ -82,18 +86,45 @@ pub struct Frame {
...
-            f,
-            "{} (method {}) -- {} ({})",
-            Address::new_id(self.source),
-            self.method,
-            &self.message,
-            self.code,
-        )
+        if let Some(data) = &self.data {
+            let data_repr = if let Ok(BytesDe(bytes)) = data.deserialize() {
+                if let Ok(s) = String::from_utf8(bytes) {
```
[the remark this answers is not in the document]
Comment (vyzo) [reply fvm/src/call_manager/backtrace.rs:97]: but why not, garbage in, garbage out :)

**Comment by vyzo** [comment]
yes it works:
```
$ ../../../target/release/fvm-bench -b ../../../../builtin-actors/output/builtin-actors.car revert.bin "" ""
message failed with backtrace:
00: f0101 (method 3844450837) -- contract reverted (33) [hello]
```

File: fvm/src/call_manager/backtrace.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply fvm/src/call_manager/backtrace.rs:99]: done.

File: fvm/src/call_manager/backtrace.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vyzo) [reply fvm/src/call_manager/backtrace.rs:97]: used suggestion above, so done.

**Comment by vyzo** [comment]
$ ../../../target/release/fvm-bench -b ../../../../builtin-actors/output/builtin-actors.car revertbig.bin "" ""
message failed with backtrace:
00: f0101 (method 3844450837) -- contract reverted (33) [abcdefghijklmnopqrstuvwyxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789- ... (skipped 32 bytes) ... FGHIJKLMNOPQRSTUVWXYZ0123456789-abcdefghijklmnopqrstuvwyxyzABCDE]

ERROR: contract execution failed

**Comment by vyzo** [comment]
ok, we are ready now -- tests with fvm-bench included.

**Comment by vyzo** [comment]
I disagree with this sentiment.
Users can do whatever they want, including having ChatGPT write their contracts, but in the end they will generate bitcode and there is exactly one way to abort with a message.
That's the `revert` opcode, which is exactly what we test here.

**Comment by Stebalien** [comment]
@vyzo I need to know that this will work for actual users using the system. If it doesn't help them debug issues (because, e.g., they get nonsense errors), there's no point in merging this change.

Yes, users _can_ do something we haven't accounted for and, in that case, they won't get our help debugging their issue. I care about the common user.

**Comment by vyzo** [comment]
Upon further discussion, this is untenable: the error from revert is type encoded, prefixed by the selector, and might not even include a string at all so this exercise is futile.

Closing as WONTFIX.

**Pull request closed without merging**